test/mongoose_lazy_routing_SUITE.erl

```diff
@@ -0,0 +1,471 @@
+                      get_all_registered_iqs()),
+    ?assertEqual([], get_all_unregistered_iqs()),
+    %% removal of ?DOMAIN_1 is required for successful second
+    %% run of can_register_iq_handler_for_domain/0
+    maybe_remove_domain(?HOST_TYPE_1, ?DOMAIN_1),
+    [IQHandlerWithHostType1, IQHandlerWithHostType2, IQHandlerWithHostType3].
+
+can_unregister_iq_handler_for_domain() ->
+    meck:reset(gen_iq_component),
+    %% trying to unregister IQ handlers registered
+    %% at can_register_iq_handler_for_domain/0
+    {ok, IQHandler1} = unregister_iq_handler_for_domain(?HOST_TYPE_2, ?NAMESPACE_1,
+                                                        ?COMPONENT),
+    {ok, IQHandler2} = unregister_iq_handler_for_domain(?HOST_TYPE_1, ?NAMESPACE_1,
+                                                        ?COMPONENT),
+    {ok, IQHandler3} = unregister_iq_handler_for_domain(?HOST_TYPE_1, ?NAMESPACE_2,
+                                                        ?COMPONENT),
+    ?assertEqual([], get_all_registered_iqs()),
+    ?assertEqualLists([{?COMPONENT, ?DOMAIN_1, ?NAMESPACE_1},
+                       {?COMPONENT, ?DOMAIN_1, ?NAMESPACE_2},
+                       {?COMPONENT, ?DOMAIN_2, ?NAMESPACE_1},
+                       {?COMPONENT, ?DOMAIN_3, ?NAMESPACE_1}],
+                      get_all_unregistered_iqs()),
+    [IQHandler1, IQHandler2, IQHandler3].
+
+can_register_and_unregister_iq_handler_for_subdomain(_Config) ->
+    AddedIQHandlers = can_register_iq_handler_for_subdomain(),
+    RemovedIQHandlers = can_unregister_iq_handler_for_subdomain(),
+    ?assertEqualLists(AddedIQHandlers, RemovedIQHandlers),
+    %% try to re-register iq handlers for subdomain
+    can_register_iq_handler_for_subdomain(),
+    ok.
+
+can_register_iq_handler_for_subdomain() ->
+    meck:reset(gen_iq_component),
+    %% add 2 subdomains for ?HOST_TYPE_2 and one domain (just to ensure that domain
+    %% doesn't affect anything)
+    ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_2)),
+    ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_3)),
+    ?assertEqual(true, maybe_add_domain_or_subdomain(?DOMAIN_2)),
+    %% register IQ handler for ?HOST_TYPE_2 subdomains
+    Pattern = subdomain_pattern(?SUBDOMAIN_2),
+    IQHandlerWithHostType1 = create_iq_handler_and_register(<<"IQH1">>, ?HOST_TYPE_2,
+                                                            Pattern, ?NAMESPACE_1,
+                                                            ?COMPONENT),
+    ?assertEqualLists([{?COMPONENT, ?SUBDOMAIN_2, ?NAMESPACE_1, IQHandlerWithHostType1},
+                       {?COMPONENT, ?SUBDOMAIN_3, ?NAMESPACE_1, IQHandlerWithHostType1}],
+                      get_all_registered_iqs()),
+    ?assertEqual([], get_all_unregistered_iqs()),
+    meck:reset(gen_iq_component),
+    %% register 2 IQ handlers for ?HOST_TYPE_1 subdomains, then add one subdomain and
+    %% one domain (just to ensure that domain doesn't affect anything)
```

Comment:
I prefer tests to follow this convention:

```
can_register_iq_handler_for_subdomain() ->
       %% GIVEN idk, for me nothing is given in this case
       %% WHEN Register IQ handler for subdomain
       ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_2)), %% ONE subdomain
       %% THEN registered
       assert_subdomain_registered(?SUBDOMAIN_2).
```

If we want to test registering two subdomains, we do another case:

```
can_register_iq_handler_for_two_subdomains() ->
       %% GIVEN idk, for me nothing is given in this case
       %% WHEN Register IQ handler for subdomains
       ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_1)),
       ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_2)),
       %% THEN registered
       assert_subdomain_registered(?SUBDOMAIN_1),
       assert_subdomain_registered(?SUBDOMAIN_2).
```

If we want to test unregister:

```
can_unregister_iq_handler_for_two_subdomains() ->
       %% GIVEN two registered domains
       ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_1)),
       ?assertEqual(true, maybe_add_domain_or_subdomain(?SUBDOMAIN_2)),
       %% WHEN Register IQ handler for subdomains
       unregister_subdomain(?SUBDOMAIN_1),
       unregister_subdomain(?SUBDOMAIN_2),
       %% THEN registered
       assert_subdomain_not_registered(?SUBDOMAIN_1),
       assert_subdomain_not_registered(?SUBDOMAIN_2).
```

I am not a big fun of asserts in the given part, so:

```
can_unregister_iq_handler_for_two_subdomains() ->
       %% GIVEN two registered domains
       given_subdomains([?SUBDOMAIN_1, ?SUBDOMAIN_2]),
       %% WHEN Unegister IQ handler for subdomains
       unregister_subdomain(?SUBDOMAIN_1),
       unregister_subdomain(?SUBDOMAIN_2),
       %% THEN not registered
       assert_subdomain_not_registered(?SUBDOMAIN_1),
       assert_subdomain_not_registered(?SUBDOMAIN_2).
```

If you check more than two things in a case, it is still ok, unless test fails. After that it's hard to understand what is actually failing.
If you do something that is not described in test name - it's very hard to understand if the test is valid or not.

Overall, it's a good idea to follow `GIVEN... WHEN... THAN...` structure. 
And make helper functions for GIVEN and THAN conditions.

And generally, it's good idea to think about post conditions. They are our side effects. What does it mean subdomain not registered? It could be encapsulated inside that `assert_subdomain_not_registered` function.

Side effects dictate the name of the test case. Action (WHEN block) in the test is the secondary, but also should be present in the name.

```
%% Was can_unregister_iq_handler_for_two_subdomains
unregistered_subdomain_is_not_listed() ->
       %% GIVEN
       given_subdomains([?SUBDOMAIN_1]),
       %% WHEN
       unregister_subdomain(?SUBDOMAIN_1),
       %% THEN not registered
       assert_subdomain_not_registered(?SUBDOMAIN_1).
       
unregistered_subdomain_looses_iq_handlers() ->
       %% GIVEN
       given_subdomains([?SUBDOMAIN_1]),
       given_iq_handler(...),
       %% WHEN
       unregister_subdomain(?SUBDOMAIN_1),
       %% THEN not registered
       assert_iq_handler_not_registered(...).
```

Also, testing with more complex preconditions is cool, but harder to read. Still can be done, but in a separate test case.

```
%% Was can_unregister_iq_handler_for_two_subdomains
unregistered_subdomain_is_not_listed_when_there_is_another_domain() ->
       %% GIVEN
       given_subdomains([?SUBDOMAIN_2]),
       unregistered_subdomain_looses_iq_handlers().
```

Reply:
I agree that splitting this huge test case could help with readability.
Personally I think that naming functions `given_*` is not a good practice - I'd rather keep the names`register_subdomains`, `register_iq_handler` etc. to be explicit about what they are exactly doing.
IMO there is no need for a significant rework here, splitting could help (test sequence in a group could suffice), but it's not a must.

Reply:
`given, when, then` template doesn't fit well for the cases that test `correct setup/correct rollback` behaviour.
I will try to split `can_register_and_unregister_iq_handler_for_domain` and `can_register_and_unregister_iq_handler_for_subdomain` test case into 4 smaller